### PR TITLE
Issue 280 : Implement frontend help/guide in the UI

### PR DESCRIPTION
**Pull request by suratdas**
Implements https://github.com/Visual-Regression-Tracker/Visual-Regression-Tracker/issues/280 

https://user-images.githubusercontent.com/9042580/127669197-5c803e14-e1bd-4469-a3fc-d9d5e3eb481b.mov

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -12,6 +12,21 @@ import {
 import Router from "./Router";
 
 function App() {
+  let helpSteps = [];
+
+  const getHelpSteps = () => {
```
Comment (pashidlos) [inline src/App.jsx:17]: could be moved into `HelpContext` 
it allows to propagate needed methods into all components

File: src/components/BuildDetails.tsx [hunk 2]
```diff
@@ -22,7 +22,7 @@ const BuildDetails: React.FunctionComponent = () => {
   return (
     <Grid container direction="column">
       <Grid item>
-        <Box m={0.5}>
+        <Box m={0.5} id="build-details">
```
Comment (pashidlos) [inline src/components/BuildDetails.tsx:25]: could we move those locators as constants to reuse?

File: src/components/BuildDetails.tsx [hunk 3]
```diff
@@ -22,7 +22,7 @@ const BuildDetails: React.FunctionComponent = () => {
   return (
     <Grid container direction="column">
```
Comment (pashidlos) [inline src/components/BuildDetails.tsx:23]: would it make sense to add locators directly to root element?

File: src/components/Header.tsx [hunk 4]
```diff
@@ -79,15 +100,39 @@ const Header: FunctionComponent = () => {
...
                 <IconButton
-                  onClick={(event: React.MouseEvent<HTMLElement>) =>
-                    setMenuRef(event.currentTarget)
-                  }
+                  onClick={handleClickStart}
                 >
-                  <Avatar>{`${user?.firstName[0]}${user?.lastName[0]}`}</Avatar>
+                  <Avatar>
+                    <HelpOutline />
+                    <Joyride
```
Comment (pashidlos) [inline src/components/Header.tsx:109]: let's move this into separate component

File: src/pages/LoginPage.tsx [hunk 5]
```diff
@@ -5,16 +5,22 @@ import { useHistory, useLocation } from "react-router-dom";
...
   const history = useHistory();
   const location = useLocation<{ from: string }>();
   const { loggedIn } = useAuthState();
   const { from } = location.state || {
     from: { pathname: routes.HOME },
   };
 
+  const helpSteps = [{
+    target: "#loginform-1",
+    content: "Get default username and password from the docker run log.",
```
Comment (pashidlos) [inline src/pages/LoginPage.tsx:18]:
```suggestion
    content: "Default admin account: visual-regression-tracker@example.com / 123456. Make sure to change it's password",
```

File: src/components/Header.tsx [hunk 6]
```diff
@@ -12,17 +12,38 @@ import { Link } from "react-router-dom";
...
+import Joyride, { CallBackProps, STATUS, StoreHelpers } from 'react-joyride';
+import { HelpOutline } from "@material-ui/icons";
 
-const Header: FunctionComponent = () => {
+const Header: FunctionComponent = (props: any) => {
   const [menuRef, setMenuRef] = React.useState<null | HTMLElement>(null);
+  const [run, setRun] = React.useState(false);
   const { loggedIn, user } = useAuthState();
   const authDispatch = useAuthDispatch();
+  let helpers: StoreHelpers;
```
Comment (pashidlos) [inline src/components/Header.tsx:23]: this one is not used 
I checked it works without 
do we need it?

File: src/pages/LoginPage.tsx [hunk 7]
```diff
@@ -26,7 +32,7 @@ const LoginPage = () => {
       justify="center"
       style={{ minHeight: "60vh" }}
     >
-      <Grid item>
+      <Grid item id="loginform-1" >
```
Comment (pashidlos) [inline src/pages/LoginPage.tsx:35]: let's move locator directly to LoginForm

File: src/pages/TestVariationListPage.tsx [hunk 8]
```diff
@@ -27,8 +27,23 @@ const TestVariationListPage: React.FunctionComponent = () => {
...
+  }, {
+    target: "#reset-filter",
+    content: (<div>Only filters items are merged to the target branch.</div>),
+  }];
+
+  const { populateHelpSteps } = props;
+
   React.useEffect(() => {
     if (projectId) {
+      populateHelpSteps(helpSteps);
```
Comment (pashidlos) [inline src/pages/TestVariationListPage.tsx:46]: move this into separate `useEffect`

File: src/pages/ProjectPage.tsx [hunk 9]
```diff
@@ -13,30 +13,49 @@ const useStyles = makeStyles((theme) => ({
...
+    target: "#build-details",
+    content: (<div>Details of the currently selected build.</div>),
+  }, {
+    target: "#test-run-list",
+    content: (<div>List of all comparisons for the selected build.</div>),
+  }];
+
+  useEffect(() => {
+    props.populateHelpSteps(helpsteps);
+  });
```
Comment (pashidlos) [inline src/pages/ProjectPage.tsx:38]: missing `deps`?

**Review by pashidlos — CHANGES_REQUESTED** [review]
that could be really useful for new users
awesome idea!

File: src/components/BuildDetails.tsx [hunk 10]
```diff
@@ -18,9 +18,10 @@ const BuildDetails: React.FunctionComponent = () => {
   }
 
   const loadingAnimation = selectedBuild.isRunning && <LinearProgress />;
+  const locatorBuildDetails = "build-details";
```
Comment (pashidlos) [inline src/components/BuildDetails.tsx:21]: could you move those kind of locators into `constants/help.ts`?

File: src/components/LoginForm.tsx [hunk 11]
```diff
@@ -28,12 +28,26 @@ const LoginForm = () => {
     );
   };
 
+  const helpSteps = [
```
Comment (pashidlos) [inline src/components/LoginForm.tsx:31]: this looks like a constant as well 
could we move those together?

File: src/contexts/help.context.tsx [hunk 12]
```diff
@@ -0,0 +1,23 @@
+import { createContext } from 'react';
+import { Step } from 'react-joyride';
+
+let helpSteps: Step[] = [];
+
+let getHelpSteps = (): Step[] => {
+    const firstStep = helpSteps[0];
+    //Below line is to prevent application breaking if element is not present for any reason (e.g. if the user deletes build or if there is no data.)
+    if (firstStep && document.getElementById(firstStep.target.toString().slice(1))) {
+        helpSteps.every((e) => {
```
Comment (pashidlos) [inline src/contexts/help.context.tsx:10]: not sure if this is working as expected
`Array.prototype.every() expects a return value from arrow function  array-callback-return`

**Review by pashidlos — CHANGES_REQUESTED** [review]
could you take a look on refactoring propose for HelpContext?
https://github.com/Visual-Regression-Tracker/frontend/pull/226

I try to make it this way based on this 
https://kentcdodds.com/blog/how-to-use-react-context-effectively

File: src/components/Header.tsx (hunk 6 above)
Comment (suratdas) [reply src/components/Header.tsx]: Removed.

File: src/components/BuildDetails.tsx (hunk 10 above)
Comment (suratdas) [reply src/components/BuildDetails.tsx]: Done.

File: src/components/LoginForm.tsx (hunk 11 above)
Comment (suratdas) [reply src/components/LoginForm.tsx:38]: Done.

File: src/contexts/help.context.tsx (hunk 12 above)
Comment (suratdas) [reply src/contexts/help.context.tsx]: Thanks. Changed it to forEach()

**Review by pashidlos — APPROVED** [review]
LGTM!

**Pull request merged**